In [1]:
import tensorflow as tf
old_v = tf.logging.get_verbosity()
tf.logging.set_verbosity(tf.logging.ERROR)

<h2>Extract MNIST data</h2>
<p style="font-size:20px">You can change the option of one_hot encoding.

In [2]:
from tensorflow.examples.tutorials.mnist import input_data
#get mnist data, with one_hot encoding
mnist = input_data.read_data_sets("MNIST_data/",one_hot=True)
#suppress warnings
tf.logging.set_verbosity(old_v)

Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting MNIST_data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


<h2>Experiments</h2>

In [27]:
# hyperparameters
NUM_ITER = 100
lr = 0.01
batch_size = 64
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:100, lr:0.01, batch_size:64, n_hidden_1:200, n_hidden_2:100, test_acc:0.8411


In [26]:
# hyperparameters
NUM_ITER = 1000
lr = 0.01
batch_size = 64
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.01, batch_size:64, n_hidden_1:200, n_hidden_2:100, test_acc:0.8781


In [29]:
# hyperparameters
NUM_ITER = 1000
lr = 0.1
batch_size = 64
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.1, batch_size:64, n_hidden_1:200, n_hidden_2:100, test_acc:0.0980


In [34]:
# hyperparameters
NUM_ITER = 1000
lr = 0.01
batch_size = 128
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.01, batch_size:128, n_hidden_1:200, n_hidden_2:100, test_acc:0.8617


In [33]:
# hyperparameters
NUM_ITER = 1000
lr = 0.01
batch_size = 256
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.01, batch_size:256, n_hidden_1:200, n_hidden_2:100, test_acc:0.8598


In [35]:
# hyperparameters
NUM_ITER = 1000
lr = 0.01
batch_size = 32
num_input = 784
n_hidden_1 = 200
n_hidden_2 = 100
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l2_out = tf.add(tf.matmul(l1_out,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.01, batch_size:32, n_hidden_1:200, n_hidden_2:100, test_acc:0.8398


In [51]:
# hyperparameters
NUM_ITER = 1000
lr = 0.001
batch_size = 64
num_input = 784
n_hidden_1 = 256
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l1_out_drop = tf.layers.dropout(l1_out, rate=0.3)
    l2_out = tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.001, batch_size:64, n_hidden_1:256, n_hidden_2:256, test_acc:0.8759


In [64]:
# hyperparameters
NUM_ITER = 10000
lr = 0.00001
batch_size = 64
num_input = 784
n_hidden_1 = 256
n_hidden_2 = 256
n_hidden_3 = 128
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'W3': tf.Variable(tf.random_normal([n_hidden_2, n_hidden_3]), name='W3'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_3, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'b3': tf.Variable(tf.zeros(shape=[n_hidden_3]), name='b3'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l1_out_drop = tf.layers.dropout(l1_out, rate=0.3)
    l2_out = tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2'])
    l2_out_drop = tf.layers.dropout(l2_out, rate=0.3)
    l3_out = tf.add(tf.matmul(l2_out_drop, weights['W3']), biases['b3'])
    out = tf.add(tf.matmul(l3_out, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, n_hidden_3:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, n_hidden_3,
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:10000, lr:1e-05, batch_size:64, n_hidden_1:256, n_hidden_2:256, n_hidden_3:128, test_acc:0.8708


In [66]:
# hyperparameters
NUM_ITER = 10000
lr = 0.00001
batch_size = 64
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
n_hidden_3 = 128
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'W3': tf.Variable(tf.random_normal([n_hidden_2, n_hidden_3]), name='W3'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_3, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'b3': tf.Variable(tf.zeros(shape=[n_hidden_3]), name='b3'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l1_out_drop = tf.layers.dropout(l1_out, rate=0.8)
    l2_out = tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2'])
    l2_out_drop = tf.layers.dropout(l2_out, rate=0.8)
    l3_out = tf.add(tf.matmul(l2_out_drop, weights['W3']), biases['b3'])
    out = tf.add(tf.matmul(l3_out, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, n_hidden_3:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, n_hidden_3,
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:10000, lr:1e-05, batch_size:64, n_hidden_1:512, n_hidden_2:256, n_hidden_3:128, test_acc:0.8781


In [70]:
# hyperparameters
NUM_ITER = 1000
lr = 0.001
batch_size = 64
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.add(tf.matmul(x,weights['W1']), biases['b1'])
    l1_out_drop = tf.layers.dropout(l1_out, rate=0.3)
    l2_out = tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2'])
    out = tf.add(tf.matmul(l2_out,weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y))
optimizer = tf.train.GradientDescentOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y})
#         if i % 100 ==0:
#             acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
#             print("step "+str(i)+", Accuracy= {:.3f}".format(acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels})))

num_iter:1000, lr:0.001, batch_size:64, n_hidden_1:512, n_hidden_2:256, test_acc:0.8800


In [98]:
# hyperparameters
NUM_ITER = 10000
lr = 0.001
batch_size = 64
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}
keep_prob = tf.placeholder(tf.float32)

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.nn.tanh(tf.add(tf.matmul(x,weights['W1']), biases['b1']))
    l1_out_drop = tf.layers.dropout(l1_out, rate=keep_prob)
    l2_out = tf.nn.tanh(tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2']))
    l2_out_drop = tf.layers.dropout(l2_out, rate=keep_prob)
    out = tf.add(tf.matmul(l2_out_drop, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y)) \
       + 0.1*(tf.reduce_sum(tf.square(biases['b1'])) + tf.reduce_sum(tf.square(biases['b2'])))
learning_rate = tf.train.exponential_decay(lr, 0, 5, 0.85, staircase=True)
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y, keep_prob:dropout_prob})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y, keep_prob:1})
            print("step %d: loss=%.6f, accuracy=%.4f" % (i, l, acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels, keep_prob:1})))

step 0: loss=25.121220, accuracy=0.1406
step 1000: loss=2.751602, accuracy=0.7500
step 2000: loss=1.135792, accuracy=0.8750
step 3000: loss=1.007750, accuracy=0.8750
step 4000: loss=0.914031, accuracy=0.9062
step 5000: loss=0.324332, accuracy=0.9531
step 6000: loss=0.658583, accuracy=0.9219
step 7000: loss=0.443663, accuracy=0.9219
step 8000: loss=0.385704, accuracy=0.9531
step 9000: loss=0.087119, accuracy=0.9844
num_iter:10000, lr:0.001, batch_size:64, n_hidden_1:512, n_hidden_2:256, test_acc:0.9101


In [96]:
# hyperparameters
NUM_ITER = 20000
lr = 0.001
batch_size = 64
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}
keep_prob = tf.placeholder(tf.float32)

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.nn.tanh(tf.add(tf.matmul(x,weights['W1']), biases['b1']))
    l1_out_drop = tf.layers.dropout(l1_out, rate=keep_prob)
    l2_out = tf.nn.tanh(tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2']))
    l2_out_drop = tf.layers.dropout(l2_out, rate=keep_prob)
    out = tf.add(tf.matmul(l2_out_drop, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y)) \
       + 0.1*(tf.reduce_sum(tf.square(biases['b1'])) + tf.reduce_sum(tf.square(biases['b2'])))
learning_rate = tf.train.exponential_decay(lr, 0, 5, 0.85, staircase=True)
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y, keep_prob:dropout_prob})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y, keep_prob:1})
            print("step %d: loss=%.6f, accuracy=%.4f" % (i, l, acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels, keep_prob:1})))

step 0: loss=24.453421, accuracy=0.2031
step 1000: loss=2.870062, accuracy=0.7344
step 2000: loss=0.908688, accuracy=0.8750
step 3000: loss=1.228595, accuracy=0.8594
step 4000: loss=0.944726, accuracy=0.8594
step 5000: loss=1.357128, accuracy=0.8594
step 6000: loss=0.335763, accuracy=0.9219
step 7000: loss=0.337492, accuracy=0.9219
step 8000: loss=0.246981, accuracy=0.9531
step 9000: loss=0.839056, accuracy=0.9062
step 10000: loss=0.130664, accuracy=0.9688
step 11000: loss=0.452667, accuracy=0.9375
step 12000: loss=0.219856, accuracy=0.9844
step 13000: loss=0.086020, accuracy=0.9844
step 14000: loss=0.355748, accuracy=0.9531
step 15000: loss=0.222085, accuracy=0.9219
step 16000: loss=0.404811, accuracy=0.9375
step 17000: loss=0.477873, accuracy=0.9375
step 18000: loss=0.198111, accuracy=0.9375
step 19000: loss=0.065092, accuracy=1.0000
num_iter:20000, lr:0.001, batch_size:64, n_hidden_1:512, n_hidden_2:256, test_acc:0.9398


In [95]:
# hyperparameters
NUM_ITER = 20000
lr = 0.001
batch_size = 64
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}
keep_prob = tf.placeholder(tf.float32)

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.nn.relu(tf.add(tf.matmul(x,weights['W1']), biases['b1']))
    l1_out_drop = tf.layers.dropout(l1_out, rate=keep_prob)
    l2_out = tf.nn.relu(tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2']))
    l2_out_drop = tf.layers.dropout(l2_out, rate=keep_prob)
    out = tf.add(tf.matmul(l2_out_drop, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y)) \
       + 0.1*(tf.reduce_sum(tf.square(biases['b1'])) + tf.reduce_sum(tf.square(biases['b2'])))
learning_rate = tf.train.exponential_decay(lr, 0, 5, 0.85, staircase=True)
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y, keep_prob:dropout_prob})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y, keep_prob:1})
            print("step %d: loss=%.6f, accuracy=%.4f" % (i, l, acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels, keep_prob:1})))

step 0: loss=1852.865356, accuracy=0.0625
step 1000: loss=20.316908, accuracy=0.9375
step 2000: loss=11.822806, accuracy=0.9375
step 3000: loss=0.009381, accuracy=1.0000
step 4000: loss=6.159378, accuracy=0.9688
step 5000: loss=0.046905, accuracy=1.0000
step 6000: loss=3.808309, accuracy=0.9844
step 7000: loss=0.002173, accuracy=1.0000
step 8000: loss=1.160165, accuracy=0.9844
step 9000: loss=0.002503, accuracy=1.0000
step 10000: loss=0.003059, accuracy=1.0000
step 11000: loss=0.002898, accuracy=1.0000
step 12000: loss=0.002666, accuracy=1.0000
step 13000: loss=0.002509, accuracy=1.0000
step 14000: loss=0.002303, accuracy=1.0000
step 15000: loss=1.620258, accuracy=1.0000
step 16000: loss=0.003455, accuracy=1.0000
step 17000: loss=0.002918, accuracy=1.0000
step 18000: loss=0.561350, accuracy=1.0000
step 19000: loss=0.003304, accuracy=1.0000
num_iter:20000, lr:0.001, batch_size:64, n_hidden_1:512, n_hidden_2:256, test_acc:0.9624


In [97]:
# hyperparameters
NUM_ITER = 20000
lr = 0.001
batch_size = 256
num_input = 784
n_hidden_1 = 512
n_hidden_2 = 256
num_classes = 10

# define variables, weights and biases
tf.reset_default_graph()
X = tf.placeholder(tf.float32,[None, num_input], name='X')
Y = tf.placeholder(tf.int32,[None, num_classes], name='Y')
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]), name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]), name='Wout')
}
biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]), name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]), name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]), name='bout')
}
keep_prob = tf.placeholder(tf.float32)

# define neural nets
def neural_net(x):
    x = tf.layers.batch_normalization(x)
    l1_out = tf.nn.relu(tf.add(tf.matmul(x,weights['W1']), biases['b1']))
    l1_out_drop = tf.layers.dropout(l1_out, rate=keep_prob)
    l2_out = tf.nn.relu(tf.add(tf.matmul(l1_out_drop,weights['W2']), biases['b2']))
    l2_out_drop = tf.layers.dropout(l2_out, rate=keep_prob)
    out = tf.add(tf.matmul(l2_out_drop, weights['Wout']), biases['bout'])
    return out

logits = neural_net(X)

# define loss and optimizer
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits, labels=Y)) \
       + 0.1*(tf.reduce_sum(tf.square(biases['b1'])) + tf.reduce_sum(tf.square(biases['b2'])))
learning_rate = tf.train.exponential_decay(lr, 0, 5, 0.85, staircase=True)
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate)
train_op = optimizer.minimize(loss)

# evaluate
correct_pred = tf.equal(tf.argmax(logits,1), tf.argmax(Y,1))
accuracy = tf.reduce_mean(tf.cast(correct_pred, tf.float32))

# initialize the variables
init = tf.global_variables_initializer()

# execute training
with tf.Session() as sess:
    sess.run(init)

    for i in range(NUM_ITER):
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        _, l = sess.run([train_op, loss], feed_dict={X:batch_x, Y:batch_y, keep_prob:dropout_prob})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y, keep_prob:1})
            print("step %d: loss=%.6f, accuracy=%.4f" % (i, l, acc))

    print("num_iter:%d, lr:%s, batch_size:%d, n_hidden_1:%d, n_hidden_2:%d, test_acc:%.4f" \
          % (NUM_ITER, lr, batch_size, n_hidden_1, n_hidden_2, 
             sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels, keep_prob:1})))

step 0: loss=3047.023438, accuracy=0.1367
step 1000: loss=29.321018, accuracy=0.9219
step 2000: loss=8.281501, accuracy=0.9648
step 3000: loss=0.315967, accuracy=0.9961
step 4000: loss=0.669692, accuracy=0.9961
step 5000: loss=1.069883, accuracy=0.9922
step 6000: loss=0.409648, accuracy=0.9961
step 7000: loss=0.000954, accuracy=1.0000
step 8000: loss=1.441097, accuracy=0.9961
step 9000: loss=0.422102, accuracy=1.0000
step 10000: loss=0.669219, accuracy=0.9922
step 11000: loss=0.001116, accuracy=1.0000
step 12000: loss=0.001227, accuracy=1.0000
step 13000: loss=0.251539, accuracy=1.0000
step 14000: loss=0.768050, accuracy=1.0000
step 15000: loss=0.001082, accuracy=1.0000
step 16000: loss=0.001622, accuracy=1.0000
step 17000: loss=1.190787, accuracy=0.9961
step 18000: loss=0.150582, accuracy=1.0000
step 19000: loss=0.001090, accuracy=1.0000
num_iter:20000, lr:0.001, batch_size:256, n_hidden_1:512, n_hidden_2:256, test_acc:0.9667


<h2>Your results</h2>

The experimental results are shown in the table below. The best performance is achieved with number of iterations 20000, learning rate 0.001, batch size 256, and activation function ReLU.

| num_iter |   lr  | batch_size | n_hidden_1 | n_hidden_2 | activation | accuracy |
|:--------:|:-----:|:----------:|:----------:|:----------:|:----------:|:--------:|
|    100   |  0.01 |     64     |     200    |     100    |     --     |  0.8411  |
|   1000   |  0.01 |     64     |     200    |     100    |     --     |  0.8781  |
|   1000   | 0.001 |     64     |     512    |     256    |     --     |  0.8800  |
|   10000  | 0.001 |     64     |     512    |     256    |    Tanh    |  0.9101  |
|   20000  | 0.001 |     64     |     512    |     256    |    Tanh    |  0.9398  |
|   10000  | 0.001 |     64     |     512    |     256    |    ReLU    |  0.9530  |
|   20000  | 0.001 |     64     |     512    |     256    |    ReLU    |  0.9624  |
|   20000  | 0.001 |     256    |     512    |     256    |    ReLU    |  0.9667  |

The setting that improves most is adding the activation function. Decreasing learning rate can also improves a little bit. However, increasing hidden neurons seems improve little to the performance.